warehouse/query-core/src/main/java/datawave/query/jexl/visitors/QueryPruningVisitor.java

```diff
@@ -0,0 +1,489 @@
+        
+        return copy;
+    }
+    
+    /**
+     * Given a JexlNode determine if the query the node represents is TRUE/FALSE/UNKNOWN
+     * 
+     * @param node
+     *            non-null JexlNode
+     * @return TRUE if evaluation would always return true, FALSE if it will always return false, otherwise UNKNOWN
+     */
+    public static TRUTH_STATE getState(JexlNode node) {
+        QueryPruningVisitor visitor = new QueryPruningVisitor(false);
+        return (TRUTH_STATE) node.jjtAccept(visitor, null);
+    }
+    
+    private boolean rewrite;
+    
+    private QueryPruningVisitor(boolean rewrite) {
+        this.rewrite = rewrite;
+    }
+    
+    // base cases find a _NOFIELD_ replace it with a FALSE, otherwise unknown
+    
+    @Override
+    public Object visit(ASTEQNode node, Object data) {
+        return identifyAndReplace(node);
+    }
+    
+    @Override
+    public Object visit(ASTNENode node, Object data) {
+        return identifyAndReplace(node);
+    }
+    
+    @Override
+    public Object visit(ASTERNode node, Object data) {
+        return identifyAndReplace(node);
+    }
+    
+    @Override
+    public Object visit(ASTNRNode node, Object data) {
+        return identifyAndReplace(node);
+    }
+    
+    @Override
+    public Object visit(ASTGENode node, Object data) {
+        return identifyAndReplace(node);
+    }
+    
+    @Override
+    public Object visit(ASTGTNode node, Object data) {
+        return identifyAndReplace(node);
+    }
+    
+    @Override
+    public Object visit(ASTLENode node, Object data) {
+        return identifyAndReplace(node);
+    }
+    
+    @Override
+    public Object visit(ASTLTNode node, Object data) {
+        return identifyAndReplace(node);
+    }
+    
+    // handle merge ups and replacements
+    
+    @Override
+    public Object visit(ASTNotNode node, Object data) {
+        if (node.jjtGetNumChildren() != 1) {
+            throw new IllegalStateException("ASTNotNode must only have one child: " + node);
+        }
+        // grab the node string before recursion so the original is intact
+        String originalString = JexlStringBuildingVisitor.buildQuery(node);
+        
+        TRUTH_STATE state = (TRUTH_STATE) node.jjtGetChild(0).jjtAccept(this, data);
+        
+        if (state == TRUTH_STATE.TRUE) {
+            replaceAndAssign(node, originalString, new ASTFalseNode(ParserTreeConstants.JJTFALSENODE));
+            return TRUTH_STATE.FALSE;
+        } else if (state == TRUTH_STATE.FALSE) {
+            replaceAndAssign(node, originalString, new ASTTrueNode(ParserTreeConstants.JJTTRUENODE));
+            return TRUTH_STATE.TRUE;
+        } else {
+            return TRUTH_STATE.UNKNOWN;
+        }
+    }
+    
+    @Override
+    public Object visit(ASTOrNode node, Object data) {
+        // grab the node string before recursion so the original is intact
+        String originalString = JexlStringBuildingVisitor.buildQuery(node);
```

Comment:
It's too bad that we can't somehow amortize the cost of capturing the `originalString` - I suspect is is expensive, and we may end up doing it in cases where we don't need to -- on a recursive basis no less.

Reply:
We use JexlStringBuildingVisitor liberally throughout the planner code. If it has a major performance problem we should consider fixing that directly. I have updated visit(And/Or node) to only do this step if debug pruning is requested, which will be a configurable option